### PR TITLE
Ekubo ETH/wstETH investment analysis

### LINKED ISSUE #1: Analyse Ekubo ETH/wstETH pool and suggest optimum LP range strategy
## Context
1. We plan to create a strkfarm managed ETH/wstETH strategy on Ekubo. We need data analysis of past data to suggest LP parameters to manage investment with max profit
2. End usage: We will be providing certain liquidity to this pool in the suggested range and we will also rebalance the position from time to time to maintain maximum fee output while keeping impermanent loss less

## Methodology
1. Divide total available liquidity and swaps data into two parts: 70-30. 70% being used for training and optimisation, 30% used for testing the result

## Expected output
1. Analyse the liquidity and swaps data of ETH/wstETH pool with best TVL
2. Suggest logic for configuring lower & upper ticks for adding liquidity along with conditions to trigger rebalance
3. Show strategy performance on the test data

_

> ⚠️ **Note:** **Overfitting must be avoided. Do whatever optimisations u want on training data, once u run test on test data, share the result as it is. Do not optimise on test data.**

_

### PR DESCRIPTION
**Pull request by ManvithaMolakala**
Resolves https://github.com/strkfarm/strategy-research/issues/1

### CODE REVIEW

**Comment by akiraonstarknet** [comment]
@ManvithaMolakala please use proper PR titles

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 1]
```diff
@@ -0,0 +1,1405 @@
+{
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:1]: Incorrect file name. Its wstETH, not wETH

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 2]
```diff
@@ -0,0 +1,1405 @@
...
+    "# To store valid events\n",
+    "valid_events_positions = []\n",
+    "valid_events_swap = []\n",
+    "\n",
+    "# Processing positions updated events\n",
+    "for event_no in range(len(events_response_positions_updated.events)):\n",
+    "    event_data = events_response_positions_updated.events[event_no].data\n",
+    "    \n",
+    "    # Check for the first condition\n",
+    "    if (hex(event_data[2]) == \"0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7\" and \\\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:659]: Do not hardcode address values. Please initiate constants in an appropriate cell above, and use them here.

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 3]
```diff
@@ -0,0 +1,1405 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "96078130",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def bool_to_sign(x):\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:715]: Functions like this are generic and since are meant to be utility, they are to be added in utils file.

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 4]
```diff
@@ -0,0 +1,1405 @@
...
+    "        events_response_positions_updated.events[event_no].data[4] == tick_spacing_touse):\n",
+    "            ekubo_eth_weth_liquidity_data.append(event_data)\n",
+    "\n",
+    "        \n",
+    "for event_no_swap in range(len(combined_events_response_swap)):\n",
+    "    event_data = {\n",
+    "        'block_number': combined_events_response_swap[event_no_swap].block_number,\n",
+    "        'transaction_hash': combined_events_response_swap[event_no_swap].transaction_hash,\n",
+    "        'price': 1.000001**(-combined_events_response_swap[event_no_swap].data[18] * bool_to_sign(combined_events_response_swap[event_no].data[19])),\n",
+    "        'tick_id': -combined_events_response_swap[event_no_swap].data[18] * bool_to_sign(combined_events_response_swap[event_no].data[19]),\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:747]: why is there a negative sign?

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 5]
```diff
@@ -0,0 +1,1405 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Step 1: Collect block numbers and prices from ekubo_eth_weth_swap_data\n",
+    "block_price_data = [{'block_number': event['block_number'], 'price': event['price']} for event in ekubo_eth_weth_swap_data]\n",
+    "\n",
+    "# Step 2: Convert the collected data into a pandas DataFrame\n",
+    "df = pd.DataFrame(block_price_data)\n",
+    "\n",
+    "# Step 3: Group by block number and calculate the median price for each block\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:813]: You could have just done on ticks directly (i.e. median tick value, instead of price) bcz finally for storing liquidity, u need tick range.

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 6]
```diff
@@ -0,0 +1,1405 @@
...
+    "    \n",
+    "    # Add the current block's liquidity to the cumulative state\n",
+    "    for liquidity_event in liquidity_data_for_block:\n",
+    "        tick_lower = liquidity_event['tick_lower']\n",
+    "        tick_upper = liquidity_event['tick_upper']\n",
+    "        delta_liquidity = liquidity_event['delta_liquidity']\n",
+    "        \n",
+    "        # Distribute liquidity to the relevant ticks in the cumulative state\n",
+    "        for tick in tick_array:\n",
+    "            if tick_lower <= tick <= tick_upper:\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:872]: i think it has to be tick < tick_upper (as per uniswap v3 math, ekubo could be different. u need to check(

File: Ekubo wETH-ETH liquidity pool.ipynb [hunk 7]
```diff
@@ -0,0 +1,1405 @@
...
+    "weth_symbol = 'WETH-USD'\n",
+    "\n",
+    "# Fetch historical data for both tokens over a given period (e.g., 1 month)\n",
+    "eth_data = yf.Ticker(eth_symbol).history(period=\"1mo\")\n",
+    "weth_data = yf.Ticker(weth_symbol).history(period=\"1mo\")\n",
+    "\n",
+    "# Align both dataframes by the same date index (merge them on the 'Date' column)\n",
+    "data = pd.DataFrame({\n",
+    "    'ETH Price': eth_data['Close'],\n",
+    "    'wETH Price': weth_data['Close']\n",
```
Comment (akiraonstarknet) [inline Ekubo wETH-ETH liquidity pool.ipynb:1255]: Tests are unfinished. Before finishing tests, writing strategy doesnt make sense as u will never know if u r doing right or wrong math.

Comment (akiraonstarknet) [reply Ekubo wETH-ETH liquidity pool.ipynb:1255]: Above strategy code will be reviewed after finishing tests. Look at pytest and configure tests for the project.

**Review by akiraonstarknet — CHANGES_REQUESTED** [review]
